app/src/org/odk/collect/android/views/ODKView.java

```diff
@@ -332,26 +330,17 @@ public void setFocus(Context context) {
     /**
      * Called when another activity returns information to answer this question.
      */
-    public void setBinaryData(Object answer) {
-        boolean set = false;
-        for (QuestionWidget q : widgets) {
-            if (q instanceof IBinaryWidget) {
-                if (((IBinaryWidget) q).isWaitingForBinaryData()) {
-                    ((IBinaryWidget) q).setBinaryData(answer);
-                    set = true;
-                    break;
-                }
-            }
+    public void setBinaryData(Object answer, PendingCalloutInterface pendingCalloutInterface) {
+        FormIndex questionFormIndex = pendingCalloutInterface.getPendingCalloutFormIndex();
+        if (questionFormIndex == null) {
+            return;
         }
 
-        if (!set) {
-            Log.w(TAG, "Attempting to return data to a widget or set of widgets not looking for data");
-                     
-            for (QuestionWidget q : widgets) {
-                if (q instanceof IBinaryWidget) {
-                    ((IBinaryWidget) q).setBinaryData(answer);
-                    break;
-                }
```

Comment:
It looks like that was there for if you had already gone through one full loop of the question widgets and not set binary data to any of them, then it did another loop through to just set data to all binary widgets, even if they didn't have the waiting for data flag set. Since you refactored this and removed the 2nd iteration through, it's not wrong to remove it.